File: AssetManager/UserInterface/Forms/GetCredentialsForm.vb
Imports System.Net
Imports System.Security

Public Class GetCredentialsForm
    Private MyCreds As NetworkCredential
    Private SecurePwd As New SecureString
    Public ReadOnly Property Credentials As NetworkCredential
        Get
            Return MyCreds
        End Get
    End Property
    Sub New()
        InitializeComponent()
    End Sub

    Sub New(credentialDescription As String)
        InitializeComponent()
        CredDescriptionLabel.Text = credentialDescription
    End Sub

    Private Sub Accept()
        Dim Username As String
        Username = Trim(txtUsername.Text)
        SecurePwd.MakeReadOnly()
        If Username <> "" And SecurePwd.Length > 0 Then
            MyCreds = New NetworkCredential(Username, SecurePwd, NetworkInfo.CurrentDomain)
            SecurePwd.Dispose()
            DialogResult = DialogResult.OK
        Else
            Message("Username or Password incomplete.", vbExclamation + vbOKOnly, "Missing Info", Me)
        End If
    End Sub

    Private Sub cmdAccept_Click(sender As Object, e As EventArgs) Handles cmdAccept.Click
        Accept()
    End Sub

    Private Sub txtPassword_KeyDown(sender As Object, e As KeyEventArgs) Handles txtPassword.KeyDown
        If e.KeyCode = Keys.Enter Then
            Accept()
        End If
        If e.KeyCode = Keys.Back Then
            If SecurePwd.Length > 0 Then
                SecurePwd.RemoveAt(SecurePwd.Length - 1)
            End If
        End If
    End Sub

    Private Sub txtPassword_KeyPress(sender As Object, e As KeyPressEventArgs) Handles txtPassword.KeyPress
        If e.KeyChar <> Convert.ToChar(Keys.Back) And e.KeyChar <> Convert.ToChar(Keys.Enter) Then
            SecurePwd.AppendChar(e.KeyChar)
        End If
    End Sub

    Private Sub txtPassword_KeyUp(sender As Object, e As KeyEventArgs) Handles txtPassword.KeyUp
        SetText()
    End Sub

    Private Sub SetText()
        txtPassword.Text = BlankText(SecurePwd.Length)
        txtPassword.SelectionStart = txtPassword.Text.Length + 1
    End Sub

    Private Function BlankText(length As Integer) As String
        If length > 0 Then
            Return New String(Char.Parse("*"), length)
        End If
        Return String.Empty
    End Function

End Class

File: AssetManager/UserInterface/Forms/GetCredentialsForm.cs
using AssetManager.UserInterface.CustomControls;
using System;
using System.Net;
using System.Security;
using System.Windows.Forms;

namespace AssetManager.UserInterface.Forms
{
    public partial class GetCredentialsForm : ExtendedForm
    {
        private NetworkCredential MyCreds;

        private SecureString SecurePwd = new SecureString();

        public NetworkCredential Credentials
        {
            get { return MyCreds; }
        }

        public GetCredentialsForm()
        {
            InitializeComponent();
        }

        public GetCredentialsForm(string credentialDescription)
        {
            InitializeComponent();
            CredDescriptionLabel.Text = credentialDescription;
        }

        private void Accept()
        {
            string Username = null;
            Username = txtUsername.Text.Trim();
            if (!string.IsNullOrEmpty(Username) & SecurePwd.Length > 0)
            {
                SecurePwd.MakeReadOnly();
                MyCreds = new NetworkCredential(Username, SecurePwd, NetworkInfo.CurrentDomain);
                SecurePwd.Dispose();
                DialogResult = DialogResult.OK;
            }
            else
            {
                OtherFunctions.Message("Username or Password incomplete.", (int)MessageBoxButtons.OK + (int)MessageBoxIcon.Exclamation, "Missing Info", this);
            }
        }

        private void cmdAccept_Click(object sender, EventArgs e)
        {
            Accept();
        }

        private void txtPassword_KeyDown(object sender, KeyEventArgs e)
        {
            if (e.KeyCode == Keys.Enter)
            {
                Accept();
            }
            if (e.KeyCode == Keys.Back)
            {
                if (SecurePwd.Length > 0)
                {
                    SecurePwd.RemoveAt(SecurePwd.Length - 1);
                }
            }
        }

        private void txtPassword_KeyPress(object sender, KeyPressEventArgs e)
        {
            if (e.KeyChar != Convert.ToChar(Keys.Back) & e.KeyChar != Convert.ToChar(Keys.Enter))
            {
                SecurePwd.AppendChar(e.KeyChar);
            }
        }

        private void txtPassword_KeyUp(object sender, KeyEventArgs e)
        {
            SetText();
        }

        private void SetText()
        {
            txtPassword.Text = BlankText(SecurePwd.Length);
            txtPassword.SelectionStart = txtPassword.Text.Length + 1;
        }

        private string BlankText(int length)
        {
            if (length > 0)
            {
                return new string(char.Parse("*"), length);
            }
            return string.Empty;
        }
    }
}